proxy.go

```diff
@@ -0,0 +1,292 @@
+	LocalAddr string
+	// FetchLocalAddr is used for looking up TCP address of the server.
+	// This is optional if you want to specify a dynamic TCP address based on incommig port.
+	FetchLocalAddr func(port int) (string, error)
+	// Log is a custom logger that can be used for the proxy.
+	// If not set a "tcp" logger is used.
+	Log logging.Logger
+}
+
+// Proxy proxies remote connection to local server.
+func (p *ProxyTCP) Proxy(remote net.Conn, msg *proto.ControlMessage) {
+	if msg.Protocol != proto.TCP {
+		panic("Proxy mismatch")
+	}
+
+	var log = p.log()
+
+	var port = msg.LocalPort
+	if port == 0 {
+		log.Warning("TCP proxy to port 0")
+	}
+
+	var localAddr = fmt.Sprintf("127.0.0.1:%d", port)
+	if p.LocalAddr != "" {
+		localAddr = p.LocalAddr
+	} else if p.FetchLocalAddr != nil {
+		l, err := p.FetchLocalAddr(msg.LocalPort)
+		if err != nil {
+			log.Warning("Failed to get custom local address: %s", err)
+			return
+		}
+		localAddr = l
+	}
+
+	log.Debug("Dialing local server: %q", localAddr)
+	local, err := net.DialTimeout("tcp", localAddr, defaultTimeout)
+	if err != nil {
+		log.Error("Dialing local server %q failed: %s", localAddr, err)
+		return
+	}
+
+	Join(local, remote, log)
+}
+
+func (p *ProxyTCP) log() logging.Logger {
+	if p.Log != nil {
+		return p.Log
+	}
+	return tpcLog
+}
+
+// Join copies data between local and remote connections.
+// It reads one connection and writes to the other.
+// It aims to provide a building block for custom Proxy implementations.
+func Join(local, remote net.Conn, log logging.Logger) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	transfer := func(side string, dst, src net.Conn) {
+		log.Debug("proxing %s -> %s", src.RemoteAddr(), dst.RemoteAddr())
+
+		n, err := io.Copy(dst, src)
+		if err != nil {
+			log.Error("%s: copy error: %s", side, err)
+		}
+
+		if err := src.Close(); err != nil {
+			log.Debug("%s: close error: %s", side, err)
+		}
+
+		// not for yamux streams, but for client to local server connections
+		if d, ok := dst.(*net.TCPConn); ok {
+			if err := d.CloseWrite(); err != nil {
+				log.Debug("%s: closeWrite error: %s", side, err)
+			}
+		}
+
+		wg.Done()
+		log.Debug("done proxing %s -> %s: %d bytes", src.RemoteAddr(), dst.RemoteAddr(), n)
+	}
+
+	go transfer("remote to local", local, remote)
+	go transfer("local to remote", remote, local)
+
+	wg.Wait()
+}
+
+////////////////////
+// ProxyOverwrite //
+////////////////////
+
+// ProxyOverwrite enables easy setting of different proxy functions for different
+// transport protocols. Consider the following example:
+//
+//     tunnel.ProxyOverwrite{
+//             HTTP: &tunnel.ProxyHTTP{
+//                     LocalAddr: localAddr,
+//             },
+//             WS: &MyCustomWSProxy{},
+//     }
+//
+// This code would result in a Proxy implementation that:
+//
+// * forwards all HTTP calls to localAddr
+// * uses MyCustomWSProxy for web sockets
+// * handles TCP using default implementation
+type ProxyOverwrite struct {
```

Comment:
@rjeczalik the current approach plays nice with default values, with map you possibly go to situations where the proxy is unexpectedly `nil`.

As for global default I don't know the only real usage is for NewClient so it could be in defined in client.go and not exported. This adds little value IMHO.